File: src/main/java/session/SessionManager.java
package session;

import equation.model.EquationItem;
import java.util.ArrayList;
import java.util.List;
import session.model.*;
import ui.custom.LineGroup;

import javax.xml.bind.*;
import javax.xml.namespace.QName;
import java.io.File;
import java.io.FileNotFoundException;
import ui.model.UIEditorItem;

public class SessionManager {

    private Session session;

    public Session newSession(){
        session = new Session();
        return session;
    }

    public void syncSession(ArrayList<UIEditorItem> uiEditorItems){
        session.setItems(new ArrayList<>());
        for(UIEditorItem editorItem : uiEditorItems){
            session.addItem(editorItem.getEditorItem());
        }
    }

    public Session openSession(File file) throws FileNotFoundException {
        if(file == null){
            throw new FileNotFoundException();
        }
        if(file.exists()){
            try {
                JAXBContext jaxbContext = JAXBContext.newInstance(Session.class, EditorItem.class, EditorItemLine.class, EquationItem.class, EditorItemZoom.class, EditorItemArea.class, EditorItemPosition.class);
                Unmarshaller jaxbUnmarshaller = jaxbContext.createUnmarshaller();
                Session session = (Session) jaxbUnmarshaller.unmarshal(file);
                session.setPath(file.getPath());
                this.session = session;
                return session;
            } catch (JAXBException e) {
                e.printStackTrace();
                System.err.println(e);
            }
        }
        throw new FileNotFoundException();
    }

    public boolean saveSession(){
        if(session == null || session.getPath() == null) return false;
        File file = new File(session.getPath());
        try {
            JAXBContext jaxbContext = JAXBContext.newInstance(Session.class, EditorItem.class, EditorItemLine.class, EquationItem.class, EditorItemZoom.class, EditorItemArea.class, EditorItemPosition.class);
            Marshaller jaxbMarshaller = jaxbContext.createMarshaller();
            jaxbMarshaller.setProperty(Marshaller.JAXB_FORMATTED_OUTPUT, true);
            jaxbMarshaller.marshal(session, file);
            return true;
        } catch (JAXBException e) {
            System.err.println(e);
            return false;
        }
    }

    public Session getSession() {
        return session;
    }

}

File: app/src/main/java/com/aragoj/session/SessionManager.kt
/*
 * Copyright 2019 franciscoaleixo
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.aragoj.session

import com.aragoj.equation.model.EquationItem
import com.aragoj.session.model.*
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import javax.xml.bind.JAXBContext

class SessionManager {
  private var cachedSession: Session? = null

  /**
   * Attempts to open a session in the given path.
   *
   * @throws javax.xml.bind.JAXBException
   * @throws java.io.FileNotFoundException
   */
  suspend fun openSession(path: String): Session = withContext(Dispatchers.IO) {
    val file = getExistingFile(path)
    val jaxbContext = JAXBContext.newInstance(Session::class.java, EditorItem::class.java,
        EditorItemLine::class.java, EquationItem::class.java, EditorItemZoom::class.java,
        EditorItemArea::class.java, EditorItemPosition::class.java)
    val jaxbUnmarshaller = jaxbContext.createUnmarshaller()
    val session = jaxbUnmarshaller.unmarshal(file) as Session
    session.path = file.path
    session
  }
}